upgrader.go

```diff
@@ -118,21 +119,25 @@ func newUpgrader(coordinationDir string, opts ...Option) (*Upgrader, error) {
 	return s, nil
 }
 
-func listenSock(coordinationDir string) (*net.UnixListener, error) {
-	listenpath := upgradeSockPath(coordinationDir, os.Getpid())
+func listenSock(osi osIface, coordinationDir string) (*net.UnixListener, error) {
+	listenpath := upgradeSockPath(coordinationDir, osi.Getpid())
 	return net.ListenUnix("unix", &net.UnixAddr{
 		Name: listenpath,
 		Net:  "unix",
 	})
 }
 
-func (u *Upgrader) AwaitUpgrade() error {
+var errClosed = errors.New("connection closed")
+
+func (u *Upgrader) awaitUpgrade() error {
 	for {
-		netConn, err := u.upgradeSock.Accept()
+		conn, err := u.upgradeSock.AcceptUnix()
 		if err != nil {
+			if strings.Contains(err.Error(), "use of closed network connection") {
```

Comment:
I really wish the standard library had a function to detect this.

Reply:
Same of course.